scaletest/trafficgen/run.go

```diff
@@ -0,0 +1,281 @@
+	}
+
+	go func() {
+		<-deadlineCtx.Done()
+		logger.Debug(ctx, "close agent connection", slog.F("agent_id", agentID))
+		_ = conn.Close()
+	}()
+
+	// Wrap the conn in a countReadWriter so we can monitor bytes sent/rcvd.
+	crw := countReadWriter{ReadWriter: conn, ctx: deadlineCtx}
+
+	// Create a ticker for sending data to the PTY.
+	tick := time.NewTicker(tickInterval)
+	defer tick.Stop()
+
+	// Now we begin writing random data to the pty.
+	rch := make(chan error, 1)
+	wch := make(chan error, 1)
+
+	go func() {
+		<-deadlineCtx.Done()
+		logger.Debug(ctx, "context deadline reached", slog.F("duration", time.Since(start)))
+	}()
+
+	// Read forever in the background.
+	go func() {
+		logger.Debug(ctx, "reading from agent", slog.F("agent_id", agentID))
+		rch <- drainContext(deadlineCtx, &crw)
+		logger.Debug(ctx, "done reading from agent", slog.F("agent_id", agentID))
+		conn.Close()
+		close(rch)
+	}()
+
+	// Write random data to the PTY every tick.
+	go func() {
+		logger.Debug(ctx, "writing to agent", slog.F("agent_id", agentID))
+		wch <- writeRandomData(deadlineCtx, &crw, bytesPerTick, tick.C)
+		logger.Debug(ctx, "done writing to agent", slog.F("agent_id", agentID))
+		close(wch)
+	}()
+
+	// Write until the context is canceled.
+	if wErr := <-wch; wErr != nil {
+		return xerrors.Errorf("write to pty: %w", wErr)
+	}
+	if rErr := <-rch; rErr != nil {
+		return xerrors.Errorf("read from pty: %w", rErr)
+	}
+
+	duration := time.Since(start)
+
+	logger.Info(ctx, "results",
+		slog.F("duration", duration),
+		slog.F("sent", crw.BytesWritten()),
+		slog.F("rcvd", crw.BytesRead()),
+	)
+
+	return nil
+}
+
+// Cleanup does nothing, successfully.
+func (*Runner) Cleanup(context.Context, string) error {
+	return nil
+}
+
+// drainContext drains from src until it returns io.EOF or ctx times out.
+func drainContext(ctx context.Context, src io.Reader) error {
+	errCh := make(chan error, 1)
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				_, err := io.CopyN(io.Discard, src, 1)
+				if ctx.Err() != nil {
+					return // context canceled while we were copying.
+				}
+				if err != nil {
+					errCh <- err
+					close(errCh)
+					return
+				}
+			}
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			close(done)
+			return nil
+		case err := <-errCh:
+			if err != nil {
+				if xerrors.Is(err, io.EOF) {
+					return nil
+				}
+				// It's OK if the context is canceled.
+				if xerrors.Is(err, context.DeadlineExceeded) {
+					return nil
+				}
+				return err
+			}
+		}
+	}
+}
+
+func writeRandomData(ctx context.Context, dst io.Writer, size int64, tick <-chan time.Time) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-tick:
+			payload := "#" + mustRandStr(size-1)
+			data, err := json.Marshal(codersdk.ReconnectingPTYRequest{
+				Data: payload,
+			})
+			if err != nil {
+				return err
+			}
+			if _, err := copyContext(ctx, dst, data); err != nil {
```

Comment:
for basically the same reasons as for the reader above, I don't think we need all this complexity on write.  When the context expires, we close the connection, which should get the writer unblocked.

You can also avoid marshalling into a buffer in this function by using a `json.Encoder`, which should save some copying and eek out maybe a little more performance.

Writing one byte at a time is going to be terrible for performance, making it harder for us to generate the load we'd like.